target/arm/translate-cheri.inc.c

```diff
@@ -0,0 +1,2045 @@
+        return false;
+
+    if (capabilities_enabled_exception(ctx))
+        return true;
+
+    bool pre = (a->op == 0b11);
+    bool post = (a->op == 0b01);
+    bool unpriv = (a->op == 0b10);
+    bool scale = (a->op != 0b00);
+
+    uint64_t imm = a->imm9;
+    if (scale)
+        imm <<= 4;
+    imm = SE_64(imm, scale ? 13 : 9);
+
+    return load_store_implementation(ctx, a->opc & 1, false, AS_ZERO(a->Ct),
+                                     REG_NONE, 4, 4, a->Rn, REG_NONE, imm, pre,
+                                     post, false, false, false, false, false, 0,
+                                     OPTION_NONE, 0, unpriv);
+}
+
+// imm12
+TRANS_F(LDR_STR1)
+{
+    if (capabilities_enabled_exception(ctx))
+        return true;
+
+    return load_store_implementation(ctx, a->L, false, AS_ZERO(a->Ct), REG_NONE,
+                                     4, 4, a->Rn, REG_NONE, a->imm12 << 4,
+                                     false, false, false, false, false, false,
+                                     false, 0, OPTION_NONE, 0, false);
+}
+
+TRANS_F(LDR_STR2)
+{
+    if (!(a->option_name & 0b10))
+        return false;
+
+    if (capabilities_enabled_exception(ctx))
+        return true;
+
+    return load_store_implementation(
+        ctx, a->opc == 0b01, false, AS_ZERO(a->Ct), REG_NONE, 4, 4, a->Rn,
+        AS_ZERO(a->Rm), 0, false, false, false, false, false, false, false, 0,
+        a->option_name, a->S ? 4 : 0, false);
+}
+
+TRANS_F(MRS_MSR)
+{
+    // MRS/MSR might touch the backing memory of DDC, which has its base as a
+    // global SYNC before the operation, and discard after.
+    // FIXME: Bit of a hack doing this here
+    tcg_gen_sync_i64(ddc_interposition);
+    handle_sys(ctx, 0x1 << 25, a->L, a->o0 + 2, a->op1, a->op2, a->CRn, a->CRm,
+               a->Ct);
+    tcg_gen_discard_i64(ddc_interposition);
+    return true;
+}
+
+TRANS_F(RR)
+{
+    if (a->opc > 0b01)
+        return false;
+
+    if (capabilities_enabled_exception(ctx))
+        return true;
+
+    void (*helper)(TCGv, TCGv_env, TCGv);
+
+    switch (a->opc) {
+    case 0: // RRLEN
+        helper = &gen_helper_crap;
+        break;
+    case 1: // RRMASK
+        helper = &gen_helper_cram;
+        break;
+    default:
+        g_assert_not_reached();
+    }
+
+    helper(cpu_reg(ctx, a->Rd), cpu_env, cpu_reg(ctx, a->Rn));
+    gen_lazy_cap_set_int(ctx, AS_ZERO(a->Rd));
+
+    return true;
+}
+
+TRANS_F(SCBNDS)
+{
+    if (capabilities_enabled_exception(ctx))
+        return true;
+
+    uint64_t length = a->S ? (a->imm6 << 4) : a->imm6;
+    gen_cheri_cap_cap_imm(ctx, a->Cd, a->Cn, length,
+                          &gen_helper_csetboundsexact);
+    return true;
+}
+
+TRANS_F(SC)
+{
+    if (capabilities_enabled_exception(ctx))
+        return true;
+
+    cheri_cap_cap_int_helper *helper = NULL;
+    if (a->opc == 0) {
+        helper = &gen_helper_csetbounds;
+    } else if (a->opc == 1) {
+        helper = &gen_helper_csetboundsexact;
```

Comment:
```suggestion
    if (!(a->opc & 2)) {
        cheri_cap_cap_int_helper *helper;
        if (a->opc == 0) {
            helper = &gen_helper_csetbounds;
        } else {
            helper = &gen_helper_csetboundsexact;
        }
        return gen_cheri_cap_cap_int(ctx, a->Cd, a->Cn, a->Rm, helper);
    }
```
would make the control flow easier to follow (the `if (helper)` at the end is a bit weird)